File: benchmark/comparison/tasks/30/solution.ts
type Result<T> = { tag: "Ok"; value: T } | { tag: "Err"; message: string };

function validate_age(age: number): Result<number> {
  if (age < 0) return { tag: "Err", message: "age must be non-negative" };
  if (age >= 150) return { tag: "Err", message: "age must be under 150" };
  return { tag: "Ok", value: age };
}

function validate_name(name: string): Result<string> {
  if (name === "") return { tag: "Err", message: "name must not be empty" };
  return { tag: "Ok", value: name };
}

function validate_user(name: string, age: number): Result<string> {
  const nameResult = validate_name(name);
  if (nameResult.tag === "Err") return nameResult;
  const ageResult = validate_age(age);
  if (ageResult.tag === "Err") return { tag: "Err", message: ageResult.message };
  return { tag: "Ok", value: `${nameResult.value} (age ${ageResult.value})` };
}

function formatResultNum(r: Result<number>): string {
  if (r.tag === "Ok") return `Ok(${r.value})`;
  return `Err(${r.message})`;
}

function formatResultStr(r: Result<string>): string {
  if (r.tag === "Ok") return `Ok(${r.value})`;
  return `Err(${r.message})`;
}

console.log(formatResultNum(validate_age(25)));
console.log(formatResultNum(validate_age(-1)));
console.log(formatResultNum(validate_age(150)));

console.log(formatResultStr(validate_name("Alice")));
console.log(formatResultStr(validate_name("")));

console.log(formatResultStr(validate_user("Alice", 25)));
console.log(formatResultStr(validate_user("", 25)));
console.log(formatResultStr(validate_user("Bob", -1)));

File: benchmark/comparison/tasks/30/solution.py
def validate_age(age):
    if age < 0:
        return ("Err", "age must be non-negative")
    if age >= 150:
        return ("Err", "age must be under 150")
    return ("Ok", age)

def validate_name(name):
    if name == "":
        return ("Err", "name must not be empty")
    return ("Ok", name)

def validate_user(name, age):
    name_r = validate_name(name)
    if name_r[0] == "Err":
        return name_r
    age_r = validate_age(age)
    if age_r[0] == "Err":
        return age_r
    return ("Ok", f"{name_r[1]} (age {age_r[1]})")

def fmt(result):
    if result[0] == "Ok":
        return f"Ok({result[1]})"
    return f"Err({result[1]})"

print(fmt(validate_age(25)))
print(fmt(validate_age(-1)))
print(fmt(validate_age(150)))
print(fmt(validate_name("Alice")))
print(fmt(validate_name("")))
print(fmt(validate_user("Alice", 25)))
print(fmt(validate_user("", 25)))
print(fmt(validate_user("Bob", -1)))
